contracts/lib/TimeLockUpgradeV3.sol

```diff
@@ -0,0 +1,155 @@
+/*
+    Copyright 2020 Set Labs Inc.
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+*/
+
+pragma solidity 0.5.7;
+
+import { Ownable } from "openzeppelin-solidity/contracts/ownership/Ownable.sol";
+import { SafeMath } from "openzeppelin-solidity/contracts/math/SafeMath.sol";
+
+
+/**
+ * @title TimeLockUpgradeV3
+ * @author Set Protocol
+ *
+ * The TimeLockUpgradeV3 contract contains a modifier for handling minimum time period updates
+ *
+ * CHANGELOG:
+ * - Requires that the caller is the owner
+ * - New function to allow deletion of existing timelocks
+ * - Added upgradeData to UpgradeRegistered event
+ */
+contract TimeLockUpgradeV3 is
+    Ownable
+{
+    using SafeMath for uint256;
+
+    /* ============ State Variables ============ */
+
+    // Timelock Upgrade Period in seconds
+    uint256 public timeLockPeriod;
+
+    // Mapping of maps hash of registered upgrade to its registration timestam
+    mapping(bytes32 => uint256) public timeLockedUpgrades;
+
+    /* ============ Events ============ */
+
+    event UpgradeRegistered(
+        bytes32 indexed _upgradeHash,
+        uint256 _timestamp,
+        bytes _upgradeData
+    );
+
+    event RemoveRegisteredUpgrade(
+        bytes32 indexed _upgradeHash
+    );
+
+    /* ============ Modifiers ============ */
+
+    modifier timeLockUpgradeV3(
+        address _validator,
+        function(address) internal returns (bool) _check
+    ) {
+        require(
```

Comment:
This seems a bit over-done. Instead of putting this check here, I'm thinking that V3 would just remove the owner check. And then you just have a separate modifier before the TimeLockUpgrade for the arbitrary check.

Reply:
Hmmm feels to me like doing that heightens the potential for misusage and allowing someone to call a function that they shouldn't be able to call. Knowing that TimeLockUpgradeV3 doesn't check to make sure the owner is the caller anymore is pretty specific domain knowledge. Whereas if someone has to pass in a verifying function that requires someone to more explicit about what they're doing.

Reply:
What types of validations are being put here that needs to be generalized?

Another approach is having the check be an unimplemented function that an inherited contract needs to implement